Composer/packages/client/src/pages/design/DebugPanel/TabExtensions/WatchTab/WatchTabContent.tsx

```diff
@@ -0,0 +1,326 @@
+      if (typeof accumulator === 'object') {
+        return accumulator[segment];
+      } else {
+        return undefined;
+      }
+    },
+    undefined
+  );
+  return value;
+};
+
+export const WatchTabContent: React.FC<DebugPanelTabHeaderProps> = ({ isActive }) => {
+  const currentProjectId = useRecoilValue(rootBotProjectIdSelector);
+  const rawWebChatTraffic = useRecoilValue(webChatTrafficState(currentProjectId ?? ''));
+  const [watchedVars, setWatchedVars] = useState<Record<string, string>>({});
+  const [pickerErrorMessages, setPickerErrorMessages] = useState<Record<string, string>>({});
+  const [selectedVars, setSelectedVars] = useState<IObjectWithKey[]>();
+  const [memoryVariablesPayload, setMemoryVariablesPayload] = useState<WatchDataPayload>({
+    kind: 'property',
+    data: { properties: [] },
+  });
+
+  const watchedVarsSelection = useMemo(
+    () =>
+      new Selection({
+        onSelectionChanged: () => {
+          setSelectedVars(watchedVarsSelection.getSelection());
+        },
+        selectionMode: SelectionMode.multiple,
+      }),
+    []
+  );
+
+  // get memory scope variables for the bot
+  useEffect(() => {
+    if (currentProjectId) {
+      const abortController = new AbortController();
+      (async () => {
+        try {
+          const variables = await getMemoryVariables(currentProjectId, { signal: abortController.signal });
+          setMemoryVariablesPayload({ kind: 'property', data: { properties: variables } });
+        } catch (e) {
+          // error can be due to abort
+        }
+      })();
+    }
+  }, [currentProjectId]);
+
+  const mostRecentBotState = useMemo(() => {
+    const botStateTraffic = rawWebChatTraffic.filter(
+      (t) => t.trafficType === 'activity' && t.activity.type === 'trace' && t.activity.name === 'BotState'
+    ) as ConversationActivityTrafficItem[];
+    if (botStateTraffic.length) {
+      return botStateTraffic[botStateTraffic.length - 1];
+    }
+  }, [rawWebChatTraffic]);
+
+  const onSelectPath = useCallback(
+    (variableId: string, path: string) => {
+      const watchedVar = Object.values(watchedVars).find((varPath) => varPath === path);
+      if (watchedVar) {
+        // the variable is already being watched, so display a validation error under the picker
+        setPickerErrorMessages({
+          ...pickerErrorMessages,
+          [variableId]: formatMessage('You are already watching this property.'),
+        });
+      } else {
+        setWatchedVars({
+          ...watchedVars,
+          [variableId]: path,
+        });
+        // clear any error messages for the variable
+        delete pickerErrorMessages[variableId];
+        setPickerErrorMessages({ ...pickerErrorMessages });
+      }
+    },
+    [pickerErrorMessages, watchedVars]
+  );
+
+  // we need to refresh the details list every time a new bot state comes in
+  const refreshedWatchedVars = useMemo(() => {
+    return Object.entries(watchedVars).map(([key, value]) => {
```

Comment:
use {...watchedVars} ?

Reply:
it needs to be of type `{ key, value }[]` so this won't work